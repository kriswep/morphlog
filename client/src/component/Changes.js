import React from 'react';
import { gql } from 'apollo-boost';
import { Query, Mutation } from 'react-apollo';
import styled from 'styled-components';
import { Button, Comment } from 'semantic-ui-react';

import Input from './Input';

const Author = styled.div`
  display: inline-block;
`;

const CHANGES_QUERY = gql`
  query changes($projectId: ID!) {
    changes(projectId: $projectId) {
      id
      createdAt
      text
      author {
        id
        name
      }
    }
  }
`;

const ADD_CHANGE_MUTATION = gql`
  mutation addChange($projectId: ID!, $txt: String!) {
    addChange(projectId: $projectId, text: $txt) {
      id
      text
      createdAt
      author {
        id
        name
      }
      project {
        id
      }
    }
  }
`;

const initialState = {
  text: '',
};

class Project extends React.Component {
  state = initialState;

  dispatch = e => {
    if (e.target.name && e.target.name) {
      const newState = {};
      newState[e.target.name] = e.target.value;
      this.setState(newState);
    }
  };

  addChange = mutate => {
    mutate({
      variables: {
        projectId: this.props.projectId,
        txt: this.state.text,
      },
      update: (proxy, { data: { addChange } }) => {
        // Read the data from our cache for this query.
        const data = proxy.readQuery({
          query: CHANGES_QUERY,
          variables: {
            projectId: addChange.project.id,
          },
        });

        // Add our todo from the mutation to the end.
        const newChanges = [addChange, ...data.changes];

        // Write our data back to the cache.
        proxy.writeQuery({
          query: CHANGES_QUERY,
          variables: {
            projectId: addChange.project.id,
          },
          data: { changes: newChanges },
        });
      },
      refetchQueries: [
        {
          query: CHANGES_QUERY,
          variables: {
            projectId: this.props.projectId,
          },
        },
      ],
    })
      .then(({ data }) => {
        console.log('got data', data);
      })
      .then(() => {
        this.setState(initialState);
      })
      .catch(error => {
        console.log('there was an error sending the query', error);
        this.setState(initialState);
      });
  };

  render() {
    return (
      <div>
        <Mutation mutation={ADD_CHANGE_MUTATION}>
          {(mutate, { error }) => (
            <div data-test="newChange">
              <Input
                textarea
                name="text"
                label="Change"
                value={this.state.text}
                type="text"
                placeholder="describe your change"
                onChange={this.dispatch}
              />
              <Button
                color="teal"
                onClick={e => {
                  e.preventDefault();
                  this.addChange(mutate);
                }}
              >
                Add
              </Button>
              {error && `Error!: ${error}`}
            </div>
          )}
        </Mutation>
        <Query
          query={CHANGES_QUERY}
          variables={{
            projectId:
              this.props.projectId || this.props.match.params.projectId,
          }}
        >
          {({ loading, error, data }) => {
            if (loading) return null;
            if (error) return `Error!: ${error}`;

            const { changes } = data;
            if (!changes) return null;
            return (
              <Comment.Group data-test="change">
                {changes.map(change => (
                  <Comment key={change.id}>
                    <Comment.Content>
                      <Comment.Author as={Author}>
                        {change.author.name}
                      </Comment.Author>
                      <Comment.Metadata>
                        <div>{new Date(change.createdAt).toLocaleString()}</div>
                      </Comment.Metadata>
                      <Comment.Text>{change.text}</Comment.Text>
                      <Comment.Actions>
                        <Comment.Action>Maybe Change</Comment.Action>
                      </Comment.Actions>
                    </Comment.Content>
                  </Comment>
                ))}
              </Comment.Group>
            );
          }}
        </Query>
      </div>
    );
  }
}

export default Project;
