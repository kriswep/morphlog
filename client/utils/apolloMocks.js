import fs from 'fs';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';

const buffer = fs.readFileSync('../schema/appSchema.graphql');
const typeDefs = buffer.toString();

const createClient = mocks => {
  const schema = makeExecutableSchema({ typeDefs });
  addMockFunctionsToSchema({
    schema,
    mocks,
  });

  const apolloCache = new InMemoryCache(window.__APOLLO_STATE__);

  return new ApolloClient({
    cache: apolloCache,
    link: new SchemaLink({ schema }),
  });
};

export default createClient;
