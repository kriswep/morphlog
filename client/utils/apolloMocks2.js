import fs from 'fs';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';
import { graphql } from 'graphql';
import { print } from 'graphql/language/printer';

const buffer = fs.readFileSync('../schema/appSchema.graphql');
const typeDefs = buffer.toString();

const createClient = async (mocks, query, args = {}) => {
  const queryString = print(query);
  const schema = makeExecutableSchema({
    typeDefs,
    resolverValidationOptions: {
      requireResolversForResolveType: false,
    },
  });
  addMockFunctionsToSchema({
    schema,
    mocks,
  });
  const res = await graphql(schema, queryString, null, null, args);

  return res.data;
};

export default createClient;
