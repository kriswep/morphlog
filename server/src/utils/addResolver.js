const addResolver = resolver => {
  const baseResolver = resolver;
  baseResolver.addResolver = childResolver => {
    const newResolver = async (parent, args, context, info) => {
      await resolver(parent, args, context, info);
      return childResolver(parent, args, context, info);
    };
    return addResolver(newResolver);
  };
  return baseResolver;
};

export default addResolver;
