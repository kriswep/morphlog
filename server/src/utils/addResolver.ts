const addResolver = resolver => {
  const baseResolver = resolver;
  baseResolver.addResolver = childResolver => {
    const newResolver = async (parent, args, context, info) => {
      await resolver(parent, args, context, info);
      return childResolver(parent, args, context, info);
    };
    return addResolver(newResolver);
  };
  return baseResolver;
};

export default addResolver;

// export interface Resolver {
//   (parent: Object, args: Object, context: Object, info: Object): Promise<
//     Resolver
//   >;
//   addResolver?: Function;
// }

// const addResolver = (resolver: Resolver): Resolver => {
//   const baseResolver = resolver;
//   baseResolver.addResolver = childResolver => {
//     const newResolver = async (
//       parent: Object,
//       args: Object,
//       context: Object,
//       info: Object,
//     ) => {
//       await resolver(parent, args, context, info);
//       return childResolver(parent, args, context, info);
//     };
//     return addResolver(newResolver);
//   };
//   return baseResolver;
// };

// export default addResolver;
