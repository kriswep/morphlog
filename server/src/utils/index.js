import * as jwt from 'jsonwebtoken';

export class AuthError extends Error {
  constructor() {
    super('Not Authenticated');
  }
}

export class AccessError extends Error {
  constructor() {
    super('No Access');
  }
}

export async function getUserId(ctx) {
  if (ctx.user && ctx.user.id) {
    return ctx.user.id;
  }
  const Authorization = ctx.request.get('Authorization');
  if (Authorization) {
    const token = Authorization.replace('Bearer ', '');
    try {
      const { userId } = jwt.verify(token, process.env.APP_SECRET);
      // check, if user is (still) in the db
      await ctx.db.exists.User({
        id: userId,
      });
      // set userId in context and return
      ctx.user = { id: userId };
      return userId;
    } catch (e) {
      throw new AuthError();
    }
  }

  throw new AuthError();
}

export async function isUserProjectAllowed(ctx, projectId) {
  const userId = await getUserId(ctx);
  if (!userId || !projectId) {
    throw new AccessError();
  }
  const projectMember = ctx.db.exists.Project({
    id: projectId,
    member_some: { id: userId },
  });
  const projectAdmin = ctx.db.exists.Project({
    id: projectId,
    admin_some: { id: userId },
  });
  if ((await projectMember) || (await projectAdmin)) {
    return true;
  }

  throw new AccessError();
}

export async function hasTeamRead(ctx, teamId) {
  const userId = await getUserId(ctx);
  if (!userId || !teamId) {
    throw new AccessError();
  }
  const hasAcess = ctx.db.exists.Team({
    id: teamId,
    OR: [
      { owner: { id: userId } },
      { admin_some: { id: userId } },
      { member_some: { id: userId } },
    ],
  });
  if (await hasAcess) {
    return true;
  }

  throw new AccessError();
}

export async function hasTeamWrite(ctx, teamId) {
  const userId = await getUserId(ctx);
  if (!userId || !teamId) {
    throw new AccessError();
  }
  const hasAcess = ctx.db.exists.Team({
    id: teamId,
    OR: [{ owner: { id: userId } }, { admin_some: { id: userId } }],
  });
  if (await hasAcess) {
    return true;
  }

  throw new AccessError();
}
