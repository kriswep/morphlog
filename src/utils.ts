import * as jwt from 'jsonwebtoken'
import { Prisma } from 'prisma-binding'

export interface Context {
  db: Prisma
  request: any
}

export function getUserId(ctx: Context) {
  const Authorization = ctx.request.get('Authorization')
  if (Authorization) {
    const token = Authorization.replace('Bearer ', '')
    const { userId } = jwt.verify(token, process.env.APP_SECRET) as { userId: string }
    return userId
  }

  throw new AuthError()
}

export class AuthError extends Error {
  constructor() {
    super('Not authorized')
  }
}

export async function isUserProjectAllowed(ctx: Context, projectId) {
  const userId = getUserId(ctx);
  const projectMember = ctx.db.exists.Project({
    id: projectId,
    member_some: { id: userId },
  })
  const projectAdmin = ctx.db.exists.Project({
    id: projectId,
    admin_some: { id: userId },
  })
  if (await projectMember || await projectAdmin) {
    console.log(projectMember, projectAdmin)
    return true;
  }

  throw new AccessError()
}

export class AccessError extends Error {
  constructor() {
    super('No Access')
  }
}
