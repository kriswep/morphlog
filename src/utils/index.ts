import * as jwt from 'jsonwebtoken';
import { Prisma, ID_Input } from '../generated/prisma';

export interface Context {
  db: Prisma;
  request: any;
  user: { id: ID_Input };
}

export class AuthError extends Error {
  constructor() {
    super('Not Authenticated');
  }
}

export class AccessError extends Error {
  constructor() {
    super('No Access');
  }
}

export function getUserId(ctx: Context): ID_Input {
  const Authorization = ctx.request.get('Authorization');
  if (Authorization) {
    const token = Authorization.replace('Bearer ', '');
    try {
      const { userId } = jwt.verify(token, process.env.APP_SECRET) as {
        userId: string;
      };
      return userId;
    } catch (e) {
      throw new AuthError();
    }
  }

  throw new AuthError();
}

export async function isUserProjectAllowed(ctx: Context, projectId) {
  const userId = getUserId(ctx);
  if (!userId || !projectId) {
    throw new AccessError();
  }
  const projectMember = ctx.db.exists.Project({
    id: projectId,
    member_some: { id: userId },
  });
  const projectAdmin = ctx.db.exists.Project({
    id: projectId,
    admin_some: { id: userId },
  });
  if ((await projectMember) || (await projectAdmin)) {
    return true;
  }

  throw new AccessError();
}
