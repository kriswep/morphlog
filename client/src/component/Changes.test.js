/* globals afterEach test expect */
import React from 'react';
import { render, cleanup, waitForElement } from 'react-testing-library';
import { getData, PreMockedProvider } from '../../utils/apolloMocks2';

import Changes, { CHANGES_QUERY } from './Changes';

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

const mocks = {
  Query: () => ({
    changes: () => [
      {
        id: '1234',
        text: 'MyChange',
        createdAt: '2018-01-21T20:45:06.000Z',
        updatedAt: '2018-01-21T20:45:06.000Z',
      },
      {
        id: '5678',
        text: 'MyOtherChange',
        createdAt: '2018-01-21T20:45:06.000Z',
        updatedAt: '2018-01-21T20:45:06.000Z',
      },
    ],
  }),

  // Mutation: () => ...
};

test('Changes renders correctly', async () => {
  const args = { projectId: 'cjcp94wxp025801100npb28yg' };
  const data = await getData({
    mocks,
    query: CHANGES_QUERY,
    args,
  });

  const { getByTestId } = render(
    <PreMockedProvider data={data} args={args} query={CHANGES_QUERY}>
      <Changes projectId="cjcp94wxp025801100npb28yg" />
    </PreMockedProvider>,
  );

  const change = await waitForElement(() => getByTestId('change'));
  const newChange = await waitForElement(() => getByTestId('newChange'));
  expect(change).toMatchSnapshot();
  expect(newChange).toMatchSnapshot();
});

const errorMock = {
  Query: () => ({
    changes: () => {
      throw new Error('changes error');
    },
  }),
};
test('Changes renders err', async () => {
  const args = { projectId: 'cjcp94wxp025801100npb28yg' };
  const data = await getData({
    mock: errorMock,
    query: CHANGES_QUERY,
    args,
  });

  const { getByTestId } = render(
    <PreMockedProvider data={data} args={args} query={CHANGES_QUERY}>
      <Changes projectId="cjcp94wxp025801100npb28yg" />
    </PreMockedProvider>,
  );

  const error = await waitForElement(() => getByTestId('error'));
  expect(error).toMatchSnapshot();
});
