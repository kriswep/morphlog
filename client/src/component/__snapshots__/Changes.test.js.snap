// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Changes renders correctly 1`] = `
<div
  data-testid="change"
  class="ui comments"
>
  <div
    class="comment"
  >
    <div
      class="content"
    >
      <div
        class="author sc-bdVaJa bSrzPW"
      >
        Hello World
      </div>
      <div
        class="metadata"
      >
        <div>
          2018-1-21 20:45:06
        </div>
      </div>
      <div
        class="text"
      >
        MyChange
      </div>
      <div
        class="actions"
      >
        <a
          class=""
        >
          Maybe Change
        </a>
      </div>
    </div>
  </div>
  <div
    class="comment"
  >
    <div
      class="content"
    >
      <div
        class="author sc-bdVaJa bSrzPW"
      >
        Hello World
      </div>
      <div
        class="metadata"
      >
        <div>
          2018-1-21 20:45:06
        </div>
      </div>
      <div
        class="text"
      >
        MyOtherChange
      </div>
      <div
        class="actions"
      >
        <a
          class=""
        >
          Maybe Change
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Changes renders correctly 2`] = `
<form
  data-testid="newChange"
  class="ui form"
>
  <div
    class="field"
  >
    <label>
      Change
    </label>
    <textarea
      name="text"
      placeholder="describe your change"
      type="text"
      rows="3"
      style="resize: none; height: auto;"
    />
  </div>
  <button
    class="ui teal button"
    role="button"
  >
    Add
  </button>
</form>
`;

exports[`Changes renders err 1`] = `
<div
  data-testid="error"
>
  Error!: 
  {"graphQLErrors":[],"networkError":{"type":"WriteError"},"message":"Network error: Error writing result to store for query:\\n query changes($projectId: ID!) {\\n  changes(projectId: $projectId) {\\n    id\\n    createdAt\\n    text\\n    author {\\n      id\\n      name\\n    }\\n  }\\n}\\n\\nCannot read property 'changes' of null"}
</div>
`;
